internals/daemon/daemon.go

```diff
@@ -363,6 +363,27 @@ func logit(handler http.Handler) http.Handler {
 	})
 }
 
+// panicHandler opts out of the default net/http behaviour of recovering from
+// panics in ServeHTTP goroutines, so that the server isn't left in a bad or
+// deadlocked state (for example, due to a held mutex lock).
+//
+// See: https://github.com/canonical/pebble/issues/314#issuecomment-1926148064
+// Workaround from: https://github.com/golang/go/issues/16542#issuecomment-246549902
+func panicHandler(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			err := recover()
+			if err != nil {
+				buf := make([]byte, 1024*1024)
```

Comment:
1MB is plenty huge. We could pass `runtime.Stack` ever-increasing buffer sizes till it fits, like the [code for `debug.Stack` does](https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/runtime/debug/stack.go;l=21), but that seems overkill. For reference, the large MAAS goroutine stack dump in [this comment](https://github.com/canonical/pebble/issues/314#issuecomment-1918769483) is about 100KB.

Reply:
Because [`debug.Stack`](https://pkg.go.dev/runtime/debug#Stack) only includes the current goroutine, whereas [`runtime.Stack`](https://pkg.go.dev/runtime#Stack) takes an `all` parameter which we're setting to true to get the stack trace from all goroutines.

By default when Go exits due to a panic it just prints the current goroutine's stack trace (ala `debug.Stack`), but that's not particularly helpful for debugging deadlock issues like this.

Reply:
Actually, on second thoughts, the full goroutine dump probably won't help us here -- it's the panic/recover that's causing a *deadlock*, but when the panic itself occurs, there's no deadlock. So I've just gone ahead and used `debug.Stack()` directly.